// @shopify/app/dist/cli/services/deploy/bundle.js
import { zip } from '@shopify/cli-kit/node/archiver';
import { renderConcurrent } from '@shopify/cli-kit/node/ui';
import { inTemporaryDirectory, mkdirSync, touchFile } from '@shopify/cli-kit/node/fs';
import { joinPath } from '@shopify/cli-kit/node/path';
import { exec } from '@shopify/cli-kit/node/system';
export async function bundleAndBuildExtensions(options) {
    await inTemporaryDirectory(async (tmpDir) => {
        const bundleDirectory = joinPath(tmpDir, 'bundle');
        await mkdirSync(bundleDirectory);
        await touchFile(joinPath(bundleDirectory, '.shopify'));
        const javyRequired = options.app.allExtensions.some((ext) => ext.features.includes('function'));
        if (javyRequired) {
            // Force the download of the javy binary in advance to avoid later problems,
            // as it might be done multiple times in parallel. https://github.com/Shopify/cli/issues/2877
            await exec('npm', ['exec', '--', 'javy', '--version'], { cwd: options.app.directory });
        }
        await renderConcurrent({
            processes: options.app.allExtensions.map((extension) => {
                return {
                    prefix: extension.localIdentifier,
                    action: async (stdout, stderr, signal) => {
                        await extension.buildForBundle({ stderr, stdout, signal, app: options.app }, options.identifiers, bundleDirectory);
                    },
                };
            }),
            showTimestamps: false,
        });
        if (options.bundlePath) {
            await zip({
                inputDirectory: bundleDirectory,
                outputZipPath: options.bundlePath,
            });
        }
    });
}
